client/txn.go

```diff
@@ -63,10 +65,16 @@ func (ts *txnSender) Send(ctx context.Context, call proto.Call) {
 // Txn is an in-progress distributed database transaction. A Txn is not safe for
 // concurrent use by multiple goroutines.
 type Txn struct {
-	db           DB
-	wrapped      Sender
-	txn          proto.Transaction
-	haveTxnWrite bool // True if there were transactional writes
+	db      DB
+	wrapped Sender
+	txn     proto.Transaction
+	// haveTxnWrite is true as soon as the current attempt contains a write
+	// (prior to sending). This is in contrast to txn.Writing, which is set
+	// by the coordinator when the first intent has been created, and which
+	// does not reset in-between retries. As such, haveTxnWrite helps deter-
+	// mine whether it makes sense to add an EndTransaction when txn.Writing
+	// is (still) unset.
+	haveTxnWrite bool
```

Comment:
> I thought you were going to move this into the protobuf so that a prior write on a prior request gets remembered

No, that was the `Writing` flag which already exists (but doesn't fully cover all bases).

`willWrite` is also inaccurate because it's true after you've written as well.

I think the current names `haveTxnEnd` and `haveTxnWrite` are pretty accurate when applied to the `retryable`.